csharp/src/Google.Protobuf/FieldMaskTree.cs

```diff
@@ -293,6 +293,13 @@ public void Merge(IMessage source, IMessage destination, FieldMask.MergeOptions
                         field.Accessor.SetValue(destination, destinationField);
                     }
 
+                    if (sourceField == null)
+                    {
+                        // If the message field is not present in the source but is in the destination, skip recursing
+                        // so we just keep the destination value recursively.
+                        continue;
```

Comment:
I'm not sure this is the correct behavior. Suppose we have a field mask of "x.y.z", and the source has "x" as an empty message (so x.y is null) but the destination has a full x.y.z, and we have options.ReplaceMessageFields == true. Shouldn't we recurse anyway, so that we end up clearing z?

Reply:
Fixed by just setting sourceField to a new template